### ISSUE TITLE
Delete old pipeline logs after X days or Y new runs

### ISSUE THREAD

**Issue by qwerty287** [woodpecker-ci/woodpecker#1068]
### Clear and concise description of the problem

From https://codeberg.org/Codeberg-CI/feedback/issues/63:

It would be nice to have an option to delete old logs of pipeline entries. Either automatic (delete logs of succesful runs after X days or Y new runs?) or with some sort of checkbox selection and action?

### Suggested solution

An option that automatically deletes old pipeline logs

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/woodpecker-ci/woodpecker/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://woodpecker-ci.org/docs/intro).
- [X] Check that there isn't already an [issue](https://github.com/woodpecker-ci/woodpecker/issues) that request the same feature to avoid creating a duplicate.

**Comment by maxkratz** [comment]
Indeed, an option to manually delete pipeline runs (for example, after debugging the CI config on a testing branch) would be nice to have.

**Comment by anbraten** [comment]
I would like to add support for automatic deletion of old pipelines.
Therefore I would suggest the "keep max last x pipelines" approach as it seems to be covering the issue (I want to keep the database size to a minimum by deleting unused content) well and seems to be the easiest solution to implement keeping things simple. This would make special api endpoints as requested in #3506 obsolete for now.

My plan would be the following:

- add UI option to set max amount of pipelines to keep per repo (we could later on add a server max amount as well)
- everytime a new pipeline is created we simply check the amount and delete the oldest ones to get back to amount x specified

**Comment by xoxys** [comment]
I don't get it. Why not keep the API first approach? Why would you hardcode this feature instead of a reusable API?

**Comment by anbraten** [comment]
> I don't get it. Why not keep the API first approach? Why would you hardcode this feature instead of a reusable API?

How should any other users benefit from this feature, will you open-source it as a service? If the plan is to introduce it as an internal feature later on wouldn't those api endpoints be unused by the internal service anyways? I guess we don't plan to call the api endpoints from the code?

**Comment by xoxys** [comment]
Still don't get your point... As I wrote in the PR already, the API can be [called by the UI](https://github.com/woodpecker-ci/woodpecker/blob/main/web/src/lib/api/index.ts#L96-L112) (that's how the UI works at all, no?). The API can also be called (from the CLI)[https://github.com/woodpecker-ci/woodpecker/blob/main/woodpecker-go/woodpecker/client.go] (the client is using the go sdk which is just doing plain API calls). If that's not true, why do we expose APIs at all? Creating an internal only function prevents manual cleanups whenever a user wants to do it with whatever range the user want to clean up.

If you don't plan to create a re-usable API first, how will the UI implementation look at all? Nevertheless, I thought woodpecker is using an api-first approach. This means users are able to write own CLI clients or even own web frontends if they want to. That's why I don't get why you don't want a public API.

**Comment by pat-s** [comment]
Reading this thread, @xoxys reasoning makes a lot of sense to me (having API endpoints which can be called from the UI and CLI) and would also cover the intentions of @anbraten AFAICS. 

@anbraten Can you eloborate what the downsides would be in your view to the approach of @xoxys?

**Comment by anbraten** [comment]
I thought about adding an api in the beginning as well. However to have an automatic deletion / retention of pipelines this has to be integrated into some kind of routine. At first there was the idea to have it executed by a cron scheduler, but then someone in chat suggested to simply remove old pipelines each time a new pipeline is created which seems pretty smart in terms of not checking inactive repos on large instances. This would make an api for this kinda obsolete. The main question remaining from what I read is if a keep x pipelines setting or a keep pipelines not older than x days setting is more useful for the majority of users.

Keep x pipelines seems to be nice as projects often have a different amounts of activity over time. Based on our UI most users are probably looking at sth like the last 10 pipelines in the UI. So keeping the 100 latest pipeline could do the job. If there is no activity for a year or so on a project you would still have some pipelines once you come back to the project.

Keeping pipelines for x days could make more sense for legal requirements, but I haven't heard of someone that this is a thing for CI pipelines yet.

**Comment by xoxys** [comment]
You still just ignore my main points.

> If you don't plan to create a reusable API first, what should the UI implementation even look like?

Is not answered.

> This would make an api for this kinda obsolete. 

No because users should still have the ability to delete pipelines whenever they want and the amount they want. I can't understand why you try to avoid an API that at all costs. However, I have outlined and explained my points pretty clear multiple times now.

**Comment by xoxys** [comment]
> Keep x pipelines seems to be nice as projects often have a different amounts of activity over time. Based on our UI most users are probably looking at sth like the last 10 pipelines in the UI. So keeping the 100 latest pipeline could do the job. If there is no activity for a year or so on a project you would still have some pipelines once you come back to the project.

I have chosen the time bases approach for the API because it can also handle "Keep x pipelines" as well. If we prefer this for e.g. the Woodpecker UI, just do:

- get last 100 pipelines via "get pipelines api" `/repos/{repo_id}/pipelines [get]`  (exists already)
- read "create date" from last pipeline in the returned list
- call "delete pipelines" api `/repos/{repo_id}/pipelines [delete]` with `/?before=...` (was proposed in the linked PR)

If some people want to use a time-based approach, they can use the same API.

**Comment by anbraten** [comment]
> You still just ignore my main points.

> > If you don't plan to create a reusable API first, what should the UI implementation even look like?

To achieve the goal "cleanup the pipeline list / database to not fill up the disk" most user probably don't wont to create an external service / cron-job. Therefore a setting like the following which enables an automatic cleanup would be sufficient for the majority:

![image](https://github.com/woodpecker-ci/woodpecker/assets/6918444/605e604e-d8f1-44dc-9947-4451becc1e6b)

If a user still has specific needs the `/repos/{repo_id}/pipelines` endpoint could be used to receive a list of pipelines and we could consider adding a generic endpoint for deleting just a single pipeline as suggested before. For even more specific needs there is always the option to manually interact with the database directly.

**Comment by xoxys** [comment]
You make assumptions about what users want instead of letting them choose on their own. As I'm a user as well, I can tell you I don't want such a setting, and it would not help me at all. However, I'm out at this point :wave:

**Comment by zc-devs** [comment]
What a hot discussion! It's funny and sad at the same time...

TLDR: go on and implement it in your way. Who wants API/whatever will implement it on top of your service.

---

> why you try to avoid an API at all costs

I doubt it is the main reason. The main is avoiding scheduling service, I think. [It is](https://github.com/woodpecker-ci/woodpecker/issues/1068#issuecomment-2044308555) and [it was](https://github.com/woodpecker-ci/woodpecker/issues/3023). I don't understand why though...
If one has [concerns in regards HA](https://github.com/go-gitea/gitea/issues/13791), then
1. [It doesn't work](https://github.com/woodpecker-ci/helm/issues/42#issuecomment-1546775954) now anyway;
2. We can develop separate micro service as Harbor did [Job service](https://github.com/goharbor/harbor/wiki/Architecture-Overview-of-Harbor), for example.

The second one is Why one should implement functionality, which is orthogonal to what he wants/needs (requirements)?
Why orthogonal? Because current feature requires just one integer - the setting of retention policy - not low-level API.

>  would like to add support for automatic deletion of old pipelines

Note _automatic_.

> I have chosen the time bases approach ... just do ...

Who? Me? Manually? ^ _Automatic_.

> users are able to write their own CLI clients or even their own web front-ends

What a beautiful definition of an user!

 > To achieve the goal "cleanup the pipeline list / database to not fill up the disk" most user probably don't wont to create an external service / cron-job

Exactly! I'm in that bucket of users. We do not consume cool Application _Programming_ Interfaces, nor write frontends. We use GUI (CLI at worst).

> I can tell you that I do not want such a setting, and it would not help me at all

Don't want - don't use. It would not help me either, but it may help somebody.

---

Anbraten are going to write internal service to clean up logs. That service will be triggered by pipeline creation (I prefer cron, as you know:). There will be retention policy setting in GUI.
Who wants to expose that service via REST API, GRPC, CLI / whatever are welcomed, I guess.

> If a user still has specific needs the /repos/{repo_id}/pipelines endpoint could be used to receive a list of pipelines and we could consider adding a generic endpoint for deleting just a single pipeline as suggested before. For even more specific needs there is always the option to manually interact with the database directly.

But it's a different task.

**Comment by xoxys** [comment]
You missed the entire point...

- I have created a PR to expose an API to delete pipelines and it wad refused
- I have not refused to implement this feature in the UI I have not even refused the way Anbraten had suggested
- I just wanted to create the API first because thats how the Woodpecker UI and the cli works already by using the API of the woodpecker server

**Comment by zc-devs** [comment]
1. That is sad, but not related to implementation in discussion. 
2. -
3. Your API is completely different to what [current implementation](https://github.com/woodpecker-ci/woodpecker/issues/1068#issuecomment-2046753037) requires. 
4. I doubt. Are you suggesting to run [this algorithm](https://github.com/woodpecker-ci/woodpecker/issues/1068#issuecomment-2044364296) in frontend? When, considering automatic aspect? 
5. It is completely different features.

**Comment by xoxys** [comment]
Having an API to delete pipelines is a different feature than having a UI setting to do it? Why do we have an API to delete a Repo if we have a button in the UI to do it? And what is the red button "Delete Repo" doing? It calls the DeleteRepo API.

And thats exactly what Im talking about.

**Comment by xoxys** [comment]
> I doubt. Are you suggesting to run this algorithm in frontend? When, considering automatic aspect?

I agree the UI might be the wrong place to automate it.

**Comment by zc-devs** [comment]
> Having an API to delete pipelines is a different feature than having a UI setting to do it?

Having an API to delete pipelines is a different feature than having an UI to set _a retention policy / setting up a trigger_.

Also, I've just read through #3506:

> Note: These APIs are about deleting the entire pipeline, not only it logs.

And this issue title: 

> Delete old pipeline logs

> deleting the entire pipeline

Don't want to search, but there was concern about audit.

>Why do we have an API to delete a Repo if we have a button in the UI to do it? And what is the red button "Delete Repo" doing? It calls the DeleteRepo API. And thats exactly what Im talking about.

I understood it. If Anbranten wanted just button to delete pipeline/s (or logs?;), than your points absolutely right.

**Comment by xoxys** [comment]
> Having an API to delete pipelines is a different feature than having an UI to set a retention policy / setting up a trigger.

You are right.

> Don't want to search, but there was concern about audit.

If that would have been the mail problem I could understand it as well but nobody said something like this not in the PR not in this discussion.

Anyway I agree it might be better to separate these topics.

**Comment by zc-devs** [comment]
I've been thinking for a while...

Hooking into pipeline lifecycle is fine if we could decouple it. Instead of calling directly `LogsCleaningService` from `PipelineService`, we could emit event `PipelineCreatedEvent` and handle it in cleaning service.

1. `LogsCleaningService` is singleton, loads configuration (retention policy) at startup.
2. Pipeline runs. At some point it is saved to DB in `PipelineService` (should already exist, didn't look for a name in the code).
3. Right after that `PipelineCreatedEvent` is created/emitted.
4. `LogsCleaningService` catches event, gets repo ID.
5. Loads (via `PipelineService` probably) pipeline runs IDs by condition from retention policy: older than X days / X count.
6. Calls method (supplying IDs from previous step) of `PipelineService` (or maybe some dedicated to logs exists `PipelineService`) to remove log entries like `removeLogsOf(int[]{1,2,3})`.

Looking forward to implementation of this feature, cause I could apply this approach to the agents cleaning also.

**Comment by shuhaowu** [comment]
As a workaround: judging by the code on how the pipeline logs are deleted today [(1)](https://github.com/woodpecker-ci/woodpecker/blob/81b7402/server/api/pipeline.go#L677-L711) [(2)](https://github.com/woodpecker-ci/woodpecker/blob/bf3f5824d97c2e7a7fa6973671fc97c187a5add4/server/services/log/file/file.go#L102-L104). It seems that it is possible to delete the logs from an external cron job as the existing code does not do any changes on the database side. If the logs path is mounted in as a volume into the docker container, a side car container can access the same directory and periodically delete the logs by using a `find -delete` command.
